core/sql/sqlcomp/CmpSeabaseDDLroutine.cpp

```diff
@@ -995,180 +1561,293 @@ void CmpSeabaseDDL::createSeabaseRoutine(
-       *CmpCommon::diags() << DgSqlCode(-1231)
-                           << DgString0(extRoutineName);
-       deallocEHI(ehi); 
-       processReturn();
-       return;
-     }
-  }
   else if (style == COM_STYLE_JAVA_OBJ ||
            style == COM_STYLE_CPP_OBJ)
-  {
-    // validate existence of the C++ or Java class in the library
-    Int32 routineHandle = NullCliRoutineHandle;
-    NAString externalPrefix(externalPath);
-    NAString externalNameForValidation(externalName);
-    NAString containerName;
+    {
+      // validate existence of the C++ or Java class in the library
+      Int32 routineHandle = NullCliRoutineHandle;
+      NAString externalPrefix(externalPath);
+      NAString externalNameForValidation(externalName);
+      NAString containerName;
 
-    if (language == COM_LANGUAGE_C || language == COM_LANGUAGE_CPP)
-      {
-        // separate the actual DLL name from the prefix
-        char separator = '/';
-        size_t separatorPos = externalPrefix.last(separator);
+      if (language == COM_LANGUAGE_C || language == COM_LANGUAGE_CPP)
+        {
+          if( libBlobHandle[0] != '\0' )
+            {
+              NAString dummyUser;
+              NAString cachedLibName, cachedLibPath;
+              
+              if (ComGenerateUdrCachedLibName(externalPrefix,redefTime,libSchNamePart,dummyUser, cachedLibName, cachedLibPath))
+                {
+                  *CmpCommon::diags() << DgSqlCode(-1231)
+                                      << DgString0(extRoutineName);
+                  deallocEHI(ehi); 
+                  processReturn();
+                  return;
+                }
+         
+              NAString cachedFullName = cachedLibPath+"/"+cachedLibName;
+          
+              if (extractLibrary(&cliInterface,libBlobHandle, (char *)cachedFullName.data()))
+                {
+                  *CmpCommon::diags() << DgSqlCode(-1231)
+                                      << DgString0(extRoutineName);
+                  deallocEHI(ehi); 
+                  processReturn();
+                  return;
+                }
+              externalPrefix = cachedLibPath;
+              containerName = cachedLibName;
+              
+            }
+          else
+            {
+              // separate the actual DLL name from the prefix
+              char separator = '/';
+              size_t separatorPos = externalPrefix.last(separator);
+
+              if (separatorPos != NA_NPOS)
+                {
+                  containerName = externalPrefix(separatorPos+1,
+                                                 externalPrefix.length()-separatorPos-1);
+                  externalPrefix.remove(separatorPos,
+                                        externalPrefix.length()-separatorPos);
+                }
+              else
+                {
+                  // assume the entire string is a local name
+                  containerName = externalPrefix;
+                  externalPrefix = ".";
```

Comment:
This adds a period to the externalPrefix, however, if a slash exists, the period is not included.  Is this intentional?